common/src/Microsoft.Azure.IIoT.Hub.Module.Client/src/Default/DeviceClientAdapter.cs

```diff
@@ -0,0 +1,238 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.IIoT.Module.Framework.Client {
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Shared;
+    using Prometheus;
+    using Serilog;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Adapts device client to interface
+    /// </summary>
+    public sealed class DeviceClientAdapter : IClient {
+
+        /// <summary>
+        /// Whether the client is closed
+        /// </summary>
+        public bool IsClosed { get; internal set; }
+
+        /// <summary>
+        /// Create client
+        /// </summary>
+        /// <param name="client"></param>
+        /// <param name="logger"></param>
+        internal DeviceClientAdapter(DeviceClient client, ILogger logger) {
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        /// <summary>
+        /// Factory
+        /// </summary>
+        /// <param name="product"></param>
+        /// <param name="cs"></param>
+        /// <param name="deviceId"></param>
+        /// <param name="transportSetting"></param>
+        /// <param name="timeout"></param>
+        /// <param name="retry"></param>
+        /// <param name="onConnectionLost"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static async Task<IClient> CreateAsync(string product,
+            IotHubConnectionStringBuilder cs, string deviceId,
+            ITransportSettings transportSetting, TimeSpan timeout,
+            IRetryPolicy retry, Action onConnectionLost, ILogger logger) {
+            var client = Create(cs, transportSetting);
+            var adapter = new DeviceClientAdapter(client, logger);
+
+            // Configure
+            client.OperationTimeoutInMilliseconds = (uint)timeout.TotalMilliseconds;
+            client.SetConnectionStatusChangesHandler((s, r) =>
+                adapter.OnConnectionStatusChange(deviceId, onConnectionLost, logger, s, r));
+            if (retry != null) {
+                client.SetRetryPolicy(retry);
+            }
+            client.ProductInfo = product;
+
+            await client.OpenAsync();
+            return adapter;
+        }
+
+        /// <inheritdoc />
+        public async Task CloseAsync() {
+            if (IsClosed) {
+                return;
+            }
+            _client.OperationTimeoutInMilliseconds = 3000;
```

Comment:
just curios why hardcoded 3000 here?

Reply:
The idea here was to not wait longer than 3 seconds for close to complete (since it involves network protocol interaction).  I would not change it at this point it has been like this for a while..